### PR TITLE
Automatically build Cromwell, update Cromwhelm on merge to develop [BW-1211]

### PR DESCRIPTION
**Pull request by aednichols**
Input:

1. Cromwell commit [`53c4693a0508518e9a6f957fbba2b1afc7dc90b5`](https://github.com/broadinstitute/cromwell/pull/6739/commits/53c4693a0508518e9a6f957fbba2b1afc7dc90b5)
    - I ran the job on my own branch for testing, but I updated the job to point to `develop` by PR time

Outputs:

1. Commit to `main` in Cromwhelm: https://github.com/broadinstitute/cromwhelm/commit/0218cf0ccad9a7e9a74c785c2101ecb77ec56a13

2. [Image in Docker Hub](https://hub.docker.com/layers/cromwell/broadinstitute/cromwell/79-cba6c97-SNAP/images/sha256-9cd6b3e404efbd1a8610b45ae606dde056dd172e6f1a83fbae36e340fb0103ea?context=explore).

3. [Complete log output of action.](https://github.com/broadinstitute/cromwell/runs/6097432895?check_suite_focus=true)

---

The action runs on self-hosted instances created by devops (`runs-on: self-hosted`) which are 2x as powerful as Travis. The whole build takes ~7 minutes, including `sbt server/docker`. My high spec Broad laptop is just a little faster at ~5 minutes.

~There is a prequisite PR https://github.com/broadinstitute/terraform-ap-deployments/pull/616 that is very close pending a naming discussion.~ Merged.

---

Now we are cool like the other repos that deploy continuously. This is near and dear to my heart:

Screen Shot 2022-04-20 at 11 58 40 AM

### CODE REVIEW

File: src/ci/bin/test.inc.sh [hunk 1]
```diff
@@ -1100,7 +1100,6 @@ cromwell::private::login_docker() {
         docker_password="$(
             cromwell::private::vault_run read -field=password secret/dsde/cromwell/common/cromwell-dockerhub || true
         )"
-        docker login --username "${docker_username}" --password-stdin <<< "${docker_password}" || true
```
Comment (aednichols) [inline src/ci/bin/test.inc.sh:1103]: not intended, reverting (cut instead of copying)

File: .github/workflows/chart_update_on_merge.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
+name: chart-update-on-merge
+
+on:
+  push: # There is no specific "PR merged" event but pushes to protected branch `develop` would only happen during PR merges (and releases) (and Jeff)
```
[the remark this answers is not in the document]
Comment (cjllanwarne) [reply .github/workflows/chart_update_on_merge.yml:4]: If you allow manual, you can set up all the params in the github Actions tab and run as many times as you want

File: .github/workflows/chart_update_on_merge.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        repository: broadinstitute/cromwhelm
+        token: ${{ secrets.BROADBOT_GITHUB_TOKEN }} # Has to be set at checkout AND later when pushing to work
+        path: cromwhelm
+    - name: Find Cromwell short SHA
+      run: |
+        cd cromwell
+        echo "CROMWELL_SHORT_SHA=`git rev-parse --short $GITHUB_SHA`" >> $GITHUB_ENV
+    - name: Find Cromwell release number
+      run: |
+        previous_version=$(curl -X GET https://api.github.com/repos/broadinstitute/cromwell/releases/latest | jq .tag_name | xargs)
```
Comment (cjllanwarne) [inline .github/workflows/chart_update_on_merge.yml:33]: Feels potentially a bit brittle... not a huge problem now, but will we get alerts if this pattern breaks and the pushes start failing? Or will it be silent failures until we notice cromwhelm has fallen super out of date?

File: .github/workflows/chart_update_on_merge.yml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    name: Cromwhelm Chart Auto Updater
+    runs-on: self-hosted # Faster machines; see https://github.com/broadinstitute/cromwell/settings/actions/runners
+    steps:
+    - name: Clone Cromwell
+      uses: actions/checkout@v2
+      with:
+        repository: broadinstitute/cromwell
+        token: ${{ secrets.BROADBOT_GITHUB_TOKEN }} # Has to be set at checkout AND later when pushing to work
+        path: cromwell
+    - uses: olafurpg/setup-scala@v10
```
Comment (cjllanwarne) [inline .github/workflows/chart_update_on_merge.yml:18]: Is this an official thing? Do we trust it (now and into the future) to do our builds safely and securely?

File: .github/workflows/chart_update_on_merge.yml [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        docker push broadinstitute/cromwell:$CROMWELL_SNAP_VERSION
+    - name: Edit & push chart
+      env:
+        BROADBOT_GITHUB_TOKEN: ${{ secrets.BROADBOT_GITHUB_TOKEN }}
+      run: |
+        cd cromwhelm
+        git checkout main
+        ls -la
+        sed -i "s/appVersion.*/appVersion: \"$CROMWELL_SNAP_VERSION\"/" cromwell-helm/Chart.yaml
+        sed -i "s/image: broadinstitute\/cromwell.*/image: broadinstitute\/cromwell:$CROMWELL_SNAP_VERSION/" cromwell-helm/templates/cromwell.yaml
```
Comment (cjllanwarne) [inline .github/workflows/chart_update_on_merge.yml:57]: Maybe worth putting comments in the yaml files along the lines of "please don't mess with the formats, they're in constant use by these (link) actions?"

**Review by cjllanwarne — APPROVED** [review]
Very, VERY exciting!

File: .github/workflows/chart_update_on_merge.yml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      run: |
+        cd cromwell
+        echo "CROMWELL_SHORT_SHA=`git rev-parse --short $GITHUB_SHA`" >> $GITHUB_ENV
+    - name: Find Cromwell release number
+      run: |
+        previous_version=$(curl -X GET https://api.github.com/repos/broadinstitute/cromwell/releases/latest | jq .tag_name | xargs)
+        echo "CROMWELL_NUMBER=$((previous_version + 1))" >> $GITHUB_ENV
+    - name: Save complete image ID
+      run: |
+        echo "CROMWELL_SNAP_VERSION=`echo "$CROMWELL_NUMBER-$CROMWELL_SHORT_SHA-SNAP"`" >> $GITHUB_ENV
```
Comment (cjllanwarne) [inline .github/workflows/chart_update_on_merge.yml:37]: TOL: In a world where everything is a SNAP, do we still need to say `-SNAP`?

**Comment by cjllanwarne** [comment]
TOL2: Do we want to have an concurrency protection on this action (eg to run at most 1 docker build at a time?) to avoid awkward race conditions (or merge conflicts) in cromwhelm if two actions are competing to update the helm chart at the same time?

File: .github/workflows/chart_update_on_merge.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (geoffjentry) [reply .github/workflows/chart_update_on_merge.yml:4]: I'll have you know I never .... well, actually I still do it.

File: .github/workflows/chart_update_on_merge.yml (hunk 4 above)
Comment (aednichols) [reply .github/workflows/chart_update_on_merge.yml:18]: It is what Denis et al. gave us in the Trivy action, which I took to be an endorsement.

File: .github/workflows/chart_update_on_merge.yml (hunk 5 above)
Comment (aednichols) [reply .github/workflows/chart_update_on_merge.yml:59]: Yes, great idea!

https://github.com/broadinstitute/cromwhelm/pull/26

File: .github/workflows/chart_update_on_merge.yml (hunk 6 above)
Comment (aednichols) [reply .github/workflows/chart_update_on_merge.yml:39]: Cromwell numbered releases do not really make a splash in the chart other than a higher version number appearing. So yes, they will be called `-SNAP`.

I decided that it is not important to call the release `80-0123456` if an hour later it could be replaced by `80-1234567-SNAP` when someone merges. Therefore, the `-SNAP` is [hard-coded when constructing `CROMWELL_SNAP_VERSION`](https://github.com/broadinstitute/cromwell/pull/6739/files#diff-9df639d711d6c489134fb2df6c9b555e7aa4b472163b925053bae66670c24b25R39).

Basically, my goal was to completely decouple the chart version from the release version. But I am sticking to existing  conventions because changing conventions has a big blast area.

File: .github/workflows/chart_update_on_merge.yml [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    - name: Login to Docker Hub
+      uses: docker/login-action@v1
+      with:
+        username: dsdejenkins
+        password: ${{ secrets.DSDEJENKINS_PASSWORD }}
+    - name: Build Cromwell Docker
+      run: |
+        cd cromwell
+        sbt server/docker
+        docker push broadinstitute/cromwell:$CROMWELL_SNAP_VERSION
```
[the remark this answers is not in the document]
Comment (aednichols) [reply .github/workflows/chart_update_on_merge.yml:50]: I did the math on this. We have had [about 70 squash-merge commits to `develop` this year](https://github.com/broadinstitute/cromwell/commits/develop).

Meanwhile, Docker Hub has the following blurb for our repo:

> There are 2793 images that haven't been pushed or pulled in over a month.

We may want to add a cleanup script in the future for those 3k images, but the order of magnitude of this particular change doesn't worry me

File: .github/workflows/chart_update_on_merge.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (aednichols) [reply .github/workflows/chart_update_on_merge.yml:5]: I saw that and thank you, but elsewhere I read that it does not distinguish between close + merge and simply closing.

[the remark this answers is not in the document]
Comment (aednichols) [reply .github/workflows/chart_update_on_merge.yml:5]: Genius! (disregard deleted comment that I made before fully reading the suggestion)

[the remark this answers is not in the document]
Comment (aednichols) [reply .github/workflows/chart_update_on_merge.yml]: Do you folks see anything wrong with [what I'm doing](https://github.com/broadinstitute/cromwell/pull/6739/files#diff-9df639d711d6c489134fb2df6c9b555e7aa4b472163b925053bae66670c24b25R4-R9) compared to [Scala Steward](https://github.com/broadinstitute/cromwell/blob/develop/.github/workflows/combine_scalasteward_prs.yml#L7-L20)?

I cannot for the life of me get the manual-run button to appear. Wondering whether it has to be checked in to a specific branch like `develop` or `master`.

File: .github/workflows/chart_update_on_merge.yml (hunk 3 above)
Comment (aednichols) [reply .github/workflows/chart_update_on_merge.yml:46]: Robustified with a regex check that exits on failure

**Comment by aednichols** [comment]
> Do we want to have an concurrency protection on this action (eg to run at most 1 docker build at a time?) to avoid awkward race conditions (or merge conflicts) in cromwhelm if two actions are competing to update the helm chart at the same time?

I have been exercising this condition fairly regularly over the last few days while pushing changes and it doesn't seem to cause any problems. The build takes a consistent 6 minutes; if I push 3 changes at 1 minute increments, the builds run on parallel nodes and finish in the order they started.

**Comment by aednichols** [comment]
Re-requesting review from @salonishah11 and @cjllanwarne. I squashed the commits and resolved finished conversations to make things manageable.

**Review by salonishah11 — APPROVED** [review]

**Review by kpierre13 — APPROVED** [review]
